core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -2313,7 +2326,7 @@ private[spark] class DAGScheduler(
     // Register merge statuses if the stage is still running and shuffle merge is not finalized yet.
     // TODO: SPARK-35549: Currently merge statuses results which come after shuffle merge
     // TODO: is finalized is not registered.
-    if (runningStages.contains(stage) && !stage.shuffleDep.shuffleMergeFinalized) {
+    if (runningStages.contains(stage) && !stage.shuffleDep.isShuffleMergeFinalizedMarked) {
```

Comment:
QQ:
We are not checking if merge id is the same here.
Consider case where `finalizeShuffleMerge` takes 10s or so - and we have stage cancellation/reexecution in the mean time.

We should do this for both `RegisterMergeStatuses` and `ShuffleMergeFinalized`
Thoughts ? (We can file a follow up jira for this if valid)

Reply:
@mridulm If `shuffleMergeId` is not same we are not finalizing the merge right? But there could be a `finalize` request for the correct `shuffleMergeId` but we could have messed up the `MergeStatuses` in that case coming from a `shuffleMergeId`. Ok makes sense. Will file a separate JIRA for this one as well.

On a side note, this is also another reason why we can `setShuffleMergeAllowed(false)` for determinate stages in `handleShuffleMergeFinalized` because it is possible that we might not finalize due to different `shuffleMergeId` finalize request.

Reply:
Agree with your analysis on both @venkata91

Reply:
Filed one for this issue - https://issues.apache.org/jira/browse/SPARK-38092